modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java

```diff
@@ -1935,41 +1935,61 @@ private boolean dumpLongRunningOperations0(long timeout) {
         final IgniteDiagnosticPrepareContext diagCtx = cctx.kernalContext().cluster().diagnosticEnabled() ?
             new IgniteDiagnosticPrepareContext(cctx.localNodeId()) : null;
 
+        WarningsGroup warnings = new WarningsGroup(diagnosticLog, 10);
+
         if (tm != null) {
             for (IgniteInternalTx tx : tm.activeTransactions()) {
                 if (curTime - tx.startTime() > timeout) {
                     found = true;
 
-                    U.warn(diagnosticLog, "Found long running transaction [startTime=" + formatTime(tx.startTime()) +
-                        ", curTime=" + formatTime(curTime) + ", tx=" + tx + ']');
+                    if (warnings.canAddMessage()) {
+                        warnings.add(">>> Transaction [startTime=" + formatTime(tx.startTime()) +
+                            ", curTime=" + formatTime(curTime) + ", tx=" + tx + ']');
+                    }
+                    else
+                        warnings.incTotal();
                 }
             }
         }
 
+        warnings.flush("First %d long running transactions [total=%d]");
```

Comment:
It would be nicer if we won't reuse warnings, but create new object each time (of smallest visibility scope possible). This way you won't need cleanup code, just discard your object.

Reply:
The smallest visibility scope possible is a good approach. But wouldn't it be better to avoid multiple objects creation? Especially if they won't be used in most cases.

Reply:
Nope, this code runs once per tens of seconds, correctness and brevity is much more important here

Reply:
Done